File: src/engine/src/wall_kick_data.rs
/*
 * Copyright (c) Grigory Alfyorov. Licensed under the MIT License.
 * See the LICENCE file in the repository root for full licence text.
 */

use std::collections::HashMap;
use std::sync::Mutex;
use serde::{Deserialize, Serialize};
use once_cell::sync::Lazy;
use crate::piece::RotationMove;
use crate::primitives::Point;

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum WallKickDataMode {
    Standard
}

type WK = HashMap<RotationMove, Vec<Point>>;

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum WallKickType {
    Default, PieceI, PieceO
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WallKickData {
    default: WK,
    piece_i: WK,
    piece_o: WK,
}

pub static WALL_KICK_DATA: Lazy<Mutex<WallKickData>> = Lazy::new(|| {
    let wkd = WallKickData::new(WallKickDataMode::Standard);
    Mutex::new(wkd)
});

const WKD_TO_RIGHT: [Point; 5] = [
    Point { x: 0, y: 0 },
    Point { x: -1, y: 0 },
    Point { x: -1, y: 1 },
    Point { x: 0, y: -2 },
    Point { x: -1, y: -2 },
];

const WKD_TO_LEFT: [Point; 5] = [
    Point { x: 0, y: 0 },
    Point { x: 1, y: 0 },
    Point { x: 1, y: 1 },
    Point { x: 0, y: -2 },
    Point { x: 1, y: -2 },
];

const WKD_LEFT_TO_INIT_DEG180: [Point; 5] = [
    Point { x: 0, y: 0 },
    Point { x: -1, y: 0 },
    Point { x: -1, y: -1 },
    Point { x: 0, y: 2 },
    Point { x: -1, y: 2 },
];

impl WallKickData {
    pub fn new(mode: WallKickDataMode) -> Self {

        match mode {
            WallKickDataMode::Standard => {
                let default = HashMap::from([
                    ( RotationMove::InitToRight,   Vec::from(WKD_TO_RIGHT)),
                    ( RotationMove::RightToInit,   vec![Point::new(0, 0), Point::new(1 , 0), Point::new(1, -1),  Point::new(0, 2),  Point::new(1, 2)] ),
                    ( RotationMove::RightToDeg180, vec![Point::new(0, 0), Point::new(1 , 0), Point::new(1, -1),  Point::new(0, 1),  Point::new(0, 2)] ),
                    ( RotationMove::Deg180ToRight, Vec::from(WKD_TO_RIGHT)),
                    ( RotationMove::Deg180ToLeft,  Vec::from(WKD_TO_LEFT)),
                    ( RotationMove::LeftToDeg180,  Vec::from(WKD_LEFT_TO_INIT_DEG180)),
                    ( RotationMove::LeftToInit,    Vec::from(WKD_LEFT_TO_INIT_DEG180)),
                    ( RotationMove::InitToLeft,    Vec::from(WKD_TO_LEFT)),
                    ( RotationMove::InitToDeg180,  vec![Point::new(0, 0), Point::new(0 , 1), Point::new(0, 2)] ),
                    ( RotationMove::Deg180ToInit,  vec![Point::new(0, 0), Point::new(0 ,-1), Point::new(0, -2)] ),
                ]);

                let piece_i = HashMap::from([
                    ( RotationMove::InitToRight,  vec![ Point::new(0, 0), Point::new(-2, 0), Point::new(1, 0),  Point::new(-2, -1), Point::new(1, 2) ] ),
                    ( RotationMove::RightToInit,  vec![ Point::new(0, 0), Point::new(2, 0),  Point::new(-1, 0), Point::new(2, 1),   Point::new(-1, -2) ] ),
                    ( RotationMove::RightToDeg180, vec![ Point::new(0, 0), Point::new(-1, 0), Point::new(2, 0),  Point::new(-1, 2),  Point::new(2, -1) ] ),
                    ( RotationMove::Deg180ToRight, vec![ Point::new(0, 0), Point::new(1, 0),  Point::new(-2, 0), Point::new(1, -2),  Point::new(-2, 1) ] ),
                    ( RotationMove::Deg180ToLeft,  vec![ Point::new(0, 0), Point::new(2, 0),  Point::new(-1, 0), Point::new(2, 1),   Point::new(-1, -2) ] ),
                    ( RotationMove::LeftToDeg180,  vec![ Point::new(0, 0), Point::new(-2, 0), Point::new(1, 0),  Point::new(-2, -1), Point::new(1, 2) ] ),
                    ( RotationMove::LeftToInit,   vec![ Point::new(0, 0), Point::new(1, 0),  Point::new(-2, 0), Point::new(1, -2),  Point::new(-2, 1) ] ),
                    ( RotationMove::InitToLeft,   vec![ Point::new(0, 0), Point::new(-1, 0), Point::new(2, 0),  Point::new(-1, 2),  Point::new(2, -1) ] ),
                    ( RotationMove::InitToDeg180, vec![ Point::new(0, 0), Point::new(0, 1),  Point::new(0, 2) ] ),
                    ( RotationMove::Deg180ToInit, vec![ Point::new(0, 0), Point::new(0, -1), Point::new(0, -2) ] ),
                ]);

                let piece_o = HashMap::from([
                    ( RotationMove::InitToRight,  vec![ Point::new(0, 0) ] ),
                    ( RotationMove::RightToInit,  vec![ Point::new(0, 0) ] ),
                    ( RotationMove::RightToDeg180, vec![ Point::new(0, 0) ] ),
                    ( RotationMove::Deg180ToRight, vec![ Point::new(0, 0) ] ),
                    ( RotationMove::Deg180ToLeft,  vec![ Point::new(0, 0) ] ),
                    ( RotationMove::LeftToDeg180,  vec![ Point::new(0, 0) ] ),
                    ( RotationMove::LeftToInit,   vec![ Point::new(0, 0) ] ),
                    ( RotationMove::InitToLeft,   vec![ Point::new(0, 0) ] ),
                    ( RotationMove::InitToDeg180, vec![ Point::new(0, 0) ] ),
                    ( RotationMove::Deg180ToInit, vec![ Point::new(0, 0) ] ),
                ]);

                WallKickData { default, piece_i, piece_o }
            }
        }
    }

    pub fn get(&self, t: &WallKickType) -> &WK {
        match t {
            WallKickType::Default => self.get_default(),
            WallKickType::PieceI => self.get_piece_i(),
            WallKickType::PieceO => self.get_piece_o()
        }
    }

    pub fn get_default(&self) -> &WK {
        &self.default
    }

    pub fn get_piece_i(&self) -> &WK {
        &self.piece_i
    }

    pub fn get_piece_o(&self) -> &WK {
        &self.piece_o
    }
}

impl Default for WallKickData {
    fn default() -> Self {
        WallKickData::new(WallKickDataMode::Standard)
    }
}

File: csharp/Quader/Quader.Engine/Pieces/WallKick/WallKickData.cs
using Quader.Engine.Primitives;

namespace Quader.Engine.Pieces.WallKick;

public class WallKickData
{
    public WallKickMode Mode { get; }

    public Dictionary<RotationMove, List<Point>> Default { get; }
    public Dictionary<RotationMove, List<Point>> PieceI { get; }
    public Dictionary<RotationMove, List<Point>> PieceO { get; }

    private static readonly List<Point> WkdToRight =
    [
        new Point(x: 0, y: 0),
        new Point(x: -1, y: 0),
        new Point(x: -1, y: 1),
        new Point(x: 0, y: -2),
        new Point(x: -1, y: -2),
    ];

    private static readonly List<Point> WkdToLeft =
    [
        new Point(x: 0, y: 0),
        new Point(x: 1, y: 0),
        new Point(x: 1, y: 1),
        new Point(x: 0, y: -2),
        new Point(x: 1, y: -2),
    ];

    private static readonly List<Point> WkdLeftToInitDeg180 =
    [
        new Point(x: 0, y: 0),
        new Point(x: -1, y: 0),
        new Point(x: -1, y: -1),
        new Point(x: 0, y: 2),
        new Point(x: -1, y: 2),
    ];

    private static readonly List<Point> WkdRightToInitDeg180 =
    [
        new Point(0, 0),
        new Point(1, 0),
        new Point(1, -1),
        new Point(0, 2),
        new Point(1, 2),
    ];

    private static readonly List<Point> WkdInitToDeg180 =
    [
        new Point(0, 0),
        new Point(0, 1),
        new Point(0, 2),
    ];

    private static readonly List<Point> WkdDeg180ToInit =
    [
        new Point(0, 0),
        new Point(0, -1),
        new Point(0, -2),
    ];

    public WallKickData(WallKickMode mode = WallKickMode.Standard)
    {
        Mode = mode;

        Default = new()
        {
            { RotationMove.InitToRight, WkdToRight },
            { RotationMove.RightToInit, WkdRightToInitDeg180 },
            { RotationMove.RightToDeg180, WkdRightToInitDeg180 },
            { RotationMove.Deg180ToRight, WkdToRight },
            { RotationMove.Deg180ToLeft, WkdToLeft },
            { RotationMove.LeftToDeg180, WkdLeftToInitDeg180 },
            { RotationMove.LeftToInit, WkdLeftToInitDeg180 },
            { RotationMove.InitToLeft, WkdToLeft },
            { RotationMove.InitToDeg180, WkdInitToDeg180 },
            { RotationMove.Deg180ToInit, WkdDeg180ToInit }
        };

        PieceI = new()
        {
            {
                RotationMove.InitToRight,
                [
                    new Point(0, 0),
                    new Point(-2, 0),
                    new Point(1, 0),
                    new Point(-2, -1),
                    new Point(1, 2),
                ]
            },
            {
                RotationMove.RightToInit,
                [
                    new Point(0, 0),
                    new Point(2, 0),
                    new Point(-1, 0),
                    new Point(2, 1),
                    new Point(-1, -2),
                ]
            },
            {
                RotationMove.RightToDeg180,
                [
                    new Point(0, 0),
                    new Point(-1, 0),
                    new Point(2, 0),
                    new Point(-1, 2),
                    new Point(2, -1),
                ]
            },
            {
                RotationMove.Deg180ToRight,
                [
                    new Point(0, 0),
                    new Point(1, 0),
                    new Point(-2, 0),
                    new Point(1, -2),
                    new Point(-2, 1),
                ]
            },
            {
                RotationMove.Deg180ToLeft,
                [
                    new Point(0, 0),
                    new Point(2, 0),
                    new Point(-1, 0),
                    new Point(2, 1),
                    new Point(-1, -2),
                ]
            },
            {
                RotationMove.LeftToDeg180,
                [
                    new Point(0, 0),
                    new Point(-2, 0),
                    new Point(1, 0),
                    new Point(-2, -1),
                    new Point(1, 2),
                ]
            },
            {
                RotationMove.LeftToInit,
                [
                    new Point(0, 0),
                    new Point(1, 0),
                    new Point(-2, 0),
                    new Point(1, -2),
                    new Point(-2, 1),
                ]
            },
            {
                RotationMove.InitToLeft,
                [
                    new Point(0, 0),
                    new Point(-1, 0),
                    new Point(2, 0),
                    new Point(-1, 2),
                    new Point(2, -1),
                ]
            },
            { RotationMove.InitToDeg180, WkdInitToDeg180 },
            { RotationMove.Deg180ToInit, WkdDeg180ToInit }
        };

        PieceO = new()
        {
            { RotationMove.InitToRight, [new Point(0, 0)] },
            { RotationMove.RightToInit, [new Point(0, 0)] },
            { RotationMove.RightToDeg180, [new Point(0, 0)] },
            { RotationMove.Deg180ToRight, [new Point(0, 0)] },
            { RotationMove.Deg180ToLeft, [new Point(0, 0)] },
            { RotationMove.LeftToDeg180, [new Point(0, 0)] },
            { RotationMove.LeftToInit, [new Point(0, 0)] },
            { RotationMove.InitToLeft, [new Point(0, 0)] },
            { RotationMove.InitToDeg180, [new Point(0, 0)] },
            { RotationMove.Deg180ToInit, [new Point(0, 0)] },
        };
    }

    public Dictionary<RotationMove, List<Point>> Get(WallKickType type)
    {
        return type switch
        {
            WallKickType.Default => Default,
            WallKickType.PieceI => PieceI,
            WallKickType.PieceO => PieceO,
            _ => throw new ArgumentOutOfRangeException(nameof(type), type, null)
        };
    }
}